### ISSUE TITLE
Backporting cached_property decorator to python < 3.8 (for lazily computing properties of immutable classes)

### ISSUE THREAD

**Issue by simonbowly** [simonbowly/math-code-guidelines#8]
# TL;DR

* The [implementation of cached_property in Python3.8](https://github.com/python/cpython/blob/50e6e991781db761c496561a995541ca8d83ff87/Lib/functools.py#L1162-L1210) is pure python so you can just drop it in (remember to `from threading import RLock()` which is the only dependency).
* The thread safety implemented there may not be necessary in mathematical cpu-bound code (typically threads in Python aren't helpful for this) so we could throw away the locks used ... but I'm not sure if they cause any overhead anyway.

Should we collect stuff like this in a utilities library that's pip installable?

# Bowly's Rant

This is just the kind of thing I find interesting re: mathematical/optimisation coding vs all of the other coding discussed on the internet!!

I went looking for how the `cached_property` decorator works as included in Python 3.8 and the discussions that led up to it. Refer to:
1. The [issue on python bug tracker](https://bugs.python.org/issue21145) discussing how to implement it. Much of the discussion is related to use in libraries like `pip` and `django` (package downloads and web servers). These libraries mostly use io-bound operations (waiting for input) so can leverage threads in python to stay responsive.
2. The [implementation in Python3.8](https://github.com/python/cpython/blob/50e6e991781db761c496561a995541ca8d83ff87/Lib/functools.py#L1162-L1210). The 3.8 implementation has thread safety built in.  This is not necessary for most of our code since it is cpu-bound (threading won't save us until the GIL goes away). **Maybe we should hedge though and keep the thread-safety in case of an optimiser that does allow callbacks from multiple threads??**
3. Some [recipe code](https://wiki.python.org/moin/PythonDecoratorLibrary#Cached_Properties) from the Python Wiki for implementing this decorator. This one implements a TTL (time to live) in the cache, which is probably meant to invalidate the result of a web request after a while.  This is not relevant if it's just being used to lazily compute properties of an immutable object.
4. A [standalone package](https://github.com/pydanny/cached-property/blob/master/cached_property.py) with various very complex approaches. Does some stuff related to `asyncio` which also isn't relevant to the immutable objects use case.

The TTL, threading and asyncio features included in these various versions could probably be discarded for the kind of code we're looking at here, this is probably sufficient (I thought there would be a simpler way but it got ugly fast when I tried...):

```python

_NOT_FOUND = object()

class cached_property:
    def __init__(self, func):
        self.func = func
        self.attrname = None

    def __set_name__(self, owner, name):
        if self.attrname is None:
            self.attrname = name
        elif name != self.attrname:
            raise TypeError(
                "Cannot assign the same cached_property to two different names "
                f"({self.attrname!r} and {name!r})."
            )

    def __get__(self, instance, owner=None):
        if instance is None:
            return self
        if self.attrname is None:
            raise TypeError("Cannot use cached_property instance without calling __set_name__ on it.")
        try:
            cache = instance.__dict__
        except AttributeError:  # not all objects have __dict__ (e.g. class defines slots)
            msg = (
                f"No '__dict__' attribute on {type(instance).__name__!r} "
                f"instance to cache {self.attrname!r} property."
            )
            raise TypeError(msg) from None
        val = cache.get(self.attrname, _NOT_FOUND)
        if val is _NOT_FOUND:
            val = self.func(instance)
            try:
                cache[self.attrname] = val
            except TypeError:
                msg = (
                    f"The '__dict__' attribute on {type(instance).__name__!r} instance "
                    f"does not support item assignment for caching {self.attrname!r} property."
                )
                raise TypeError(msg) from None
        return val
```
